### PR TITLE
Navigate to partial template on GitHub from GOVUK

### PR DESCRIPTION
**Pull request by ikennaokpala**
Related PR https://github.com/alphagov/govuk-toolkit-chrome/pull/78
[Trello card](https://trello.com/c/OS0nF0tp/272-8-provide-a-way-for-content-designers-to-navigate-from-content-to-partial-templates-on-github)

## Motivation

When a content designer needs to edit smart answer content, they find it difficult to find the relevant template(s) and partial(s) that a outcome or question is comprised of.

This PR continues the spike from [this PR][1] with the aim of wrapping and tagging each partial for identification, highlighting and linking to the relevant partial template on GitHub.

An example can be found when looking at templates like [this one][2], that have nested partial templates within a content_for title.

This PR also caters for `.txt` requests by making sure that the wrapping debug div is only applied for standard HTML requests.

[1]: https://github.com/alphagov/smart-answers/pull/2642
[2]: https://www.gov.uk/marriage-abroad/y/sweden/uk/partner_british/same_sex

## Factcheck

[Preview link for html format](https://smart-answers-preview-pr-3195.herokuapp.com/marriage-abroad/y/sweden/uk/partner_british/same_sex)
[Preview link for text format](https://smart-answers-preview-pr-3195.herokuapp.com/marriage-abroad/y/sweden/uk/partner_british/same_sex.txt)

[GOVUK for html format](https://gov.uk/marriage-abroad/y/sweden/uk/partner_british/same_sex)

## Expected changes

- Content designers can now locate partial templates on Github

## Before

screen shot 2017-08-22 at 09 20 29

## After

screen shot 2017-08-23 at 09 23 28

### CODE REVIEW

**Comment by ikennaokpala** [comment]
Rebasing from 6ea29f4e67c710c61e2d053d127590ccafbee944 to 3f46068a00a0abd06b647903a09b563f036c75cb

File: lib/smart_answer_partial_template_wrapper.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  def govspeak_to_html(content)
+    Govspeak::Document.new(remove_leading_spaces(content))
+      .to_html
+      .strip
+      .html_safe
+  end
+
+  def remove_leading_spaces(content)
+    content.to_str.gsub(/^ +/, "")
+  end
```
Comment (floehopper) [inline lib/smart_answer_partial_template_wrapper.rb]: This seems to duplicate logic from `ErbRenderer#strip_leading_spaces`. Is this duplication really necessary. If so, you should explain why in the commit note. If not, it should be removed.

File: lib/smart_answer_partial_template_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+private
+
+  def valid_path_and_format?(context)
+    smart_answer_partial_path? && html_format?(context)
+  end
+
+  def html_format?(context)
+    context.request&.format&.symbol.to_s == "html"
```
Comment (floehopper) [inline lib/smart_answer_partial_template_wrapper.rb]: I think that the use of `&.` is a bit of a code "smell". I would've though that `context.request.format` would always return something and I'm not convinced it necessary to convert the format into a symbol and then into into a String in order to do the comparison, e.g. could we compare it to an instance of a `Mime::Type`?

File: lib/smart_answer_partial_template_wrapper.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    @template_identifier_file_name ||= template_identifier_path_name
+      .basename
+      .to_s
+  end
+
+  def under_smart_answer_flow_directory?
+    template_identifier_path.present? &&
+      template_identifier_path.include?("lib/smart_answer_flows")
+  end
+end
```
Comment (floehopper) [inline lib/smart_answer_partial_template_wrapper.rb]: It feels as if this module needs to "know" quite a bit about the application as a whole, e.g. the directory where Smart Answer flows are defined, the fact that Smart Answer templates have a ".govspeak.erb" file extension, etc. This seems less than ideal, especially given that this module is used to render *all* partial templates across the application.

File: test/unit/smart_answer_partial_template_wrapper_test.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        TestRenderer.any_instance.stubs(:render).returns("## Hello")
+
+        assert_match(
+          /^<div data-debug-template-path=\".*#{@path}\"><h2 id=\"hello\">Hello<\/h2><\/div>/,
+          @sample_renderer.render(@context, {}, nil)
+        )
+      end
+    end
+  end
+end
```
Comment (floehopper) [inline test/unit/smart_answer_partial_template_wrapper_test.rb]: I view the use of `any_instance` stubbing as a bit of a code "smell", because it makes the test hard to follow and reason about. I would try to re-write the test and, if necessary, the implementation under test to reduce/eliminate the use of this type of stubbing.

File: lib/smart_answer/title.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      document_fragment.children.first.attributes.first.last.value
+    end
+
+  private
+
+    def document_fragment
+      @document_fragment ||= Nokogiri::HTML::DocumentFragment.parse(@content)
+    end
+  end
+end
```
Comment (floehopper) [inline lib/smart_answer/title.rb]: The logic in this class seems to be exclusively view/presentation-related and as such I think this should be a presenter/renderer class rather than what looks to be more of a "model" class at the moment.

**Comment by ikennaokpala** [comment]
@floehopper 

> I think you should include an "Implementation" section in the PR description which summarises the overall implementation. You could probably steal some of the text from this PR description. This would also be a good place to reference the Chrome toolkit PR.



> Unless I'm missing something the final commit seems to be orthogonal to the purpose of the rest of the PR. If so, it should be made as a separate PR.

The final commit was added because on integration the  `EXPOSE_GOVSPEAK` environment variable was not set. If we conclude this we feel this should be commit should be included to allow all environments to respond .txt

> The main (2nd) commit is really quite large and consequently hard to understand all of its impact. Ideally it would be good to split this up a bit more. 

I will look into how to section this commit.

> I haven't really had time to look at the changes closely enough, but I'm concerned that knowledge about bits of the app might've leaked into places where it shouldn't be.

Thanks for raising this.. It appears the prefixes from the lookup context may have been the reason why the ern renderer did not setup with the controller initially. Hence why ? I have decided to include it. 

> Have you checked whether introducing this change has any significant negative impact on performance. Given that you are changing production behaviour, I think this is an important consideration.

The performance appears unchanged.. I will check with colleagues with access to instrumentation tools for further insight.

> Overall I'm a bit sceptical whether the benefit to content editors merits the significant increase in the complexity of the production code, particularly given that it's in some sense "debug" functionality. I think I'd want to be convinced that the extra complexity (and hence maintenance overhead) is really worthwhile. Do you have any feedback / user research from content editors to help make this judgement (I don't have access to the underlying Trello card)?

I beg to differ. I believe this presents the opportunity to further review smart answers as a whole and make the work of content editors easier.

File: lib/smart_answer/erb_renderer.rb [hunk 6]
```diff
@@ -10,11 +10,17 @@ def options(options = nil)
       end
     end
 
-    def initialize(template_directory:, template_name:, locals: {}, helpers: [])
+    def initialize(template_directory:, template_name:, locals: {}, helpers: [], options: {})
```
[the remark this answers is not in the document]
Comment (ikennaokpala) [reply lib/smart_answer/erb_renderer.rb]: In some respects this area is an experimental part of this work.. I agree with you and will refactor for controller to be a named keword argument.

File: lib/smart_answer/erb_renderer.rb [hunk 7]
```diff
@@ -10,11 +10,17 @@ def options(options = nil)
...
       @template_name = template_name
       @locals = locals
-      @view = ActionView::Base.new([@template_directory, FlowRegistry.instance.load_path])
+      controller = options[:controller]
+      @view = ActionView::Base.new(
+        [@template_directory, FlowRegistry.instance.load_path],
+        {},
+        controller
+      )
+      @view.lookup_context.prefixes = [] if valid_controller?(controller)
```
[the remark this answers is not in the document]
Comment (ikennaokpala) [reply lib/smart_answer/erb_renderer.rb]: This is me being a bit flexible with foresight in mind.. I had an eye on the a possible scenario where a controller wasn't passed.

File: test/unit/erb_renderer_test.rb [hunk 8]
```diff
@@ -250,6 +250,49 @@ class ErbRendererTest < ActiveSupport::TestCase
       end
     end
 
+    test 'can setup action view without a controller' do
```
[the remark this answers is not in the document]
Comment (ikennaokpala) [reply test/unit/erb_renderer_test.rb]: To be honest there isn't much case to be made.. unless the presenters don't supplied one

File: lib/smart_answer_partial_template_wrapper.rb (hunk 2 above)
Comment (ikennaokpala) [reply lib/smart_answer_partial_template_wrapper.rb]: This should have been removed some other options were experimented with.. and this led to this..  Once the controller is supplied the format will exist.

File: lib/smart_answer_partial_template_wrapper.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  def html_format?(context)
+    context.request&.format&.symbol.to_s == "html"
+  end
+
+  def govspeak_to_html(content)
+    Govspeak::Document.new(remove_leading_spaces(content))
+      .to_html
+      .strip
+      .html_safe
+  end
```
[the remark this answers is not in the document]
Comment (ikennaokpala) [reply lib/smart_answer_partial_template_wrapper.rb]: It is required, because the starting spaces don't get converted into  Govspeak.

File: test/unit/erb_renderer_test.rb [hunk 10]
```diff
@@ -250,6 +250,49 @@ class ErbRendererTest < ActiveSupport::TestCase
...
+
+      ErbRenderer.new(
+        template_directory: '/path',
+        template_name: 'temp-name',
+        locals: {},
+        helpers: [],
+        options: { controller: controller }
+      )
+    end
+
```
[the remark this answers is not in the document]
Comment (ikennaokpala) [reply test/unit/erb_renderer_test.rb]: They are useful as they ensure that the `erb_renderer` is setup as expected.

File: app.json [hunk 11]
```diff
@@ -8,7 +8,6 @@
     "PLEK_SERVICE_CONTENT_STORE_URI": "https://www.gov.uk/api" ,
     "PLEK_SERVICE_STATIC_URI": "https://assets-origin.integration.publishing.service.gov.uk/",
     "RUNNING_ON_HEROKU": "true",
-    "EXPOSE_GOVSPEAK": "true",
```
[the remark this answers is not in the document]
Comment (ikennaokpala) [reply app.json]: This last commit allows the us to test this on environments like integration which have since changed and no longer set EXPOSE_GOVSPEAK.

File: app/assets/javascripts/helpers.js [hunk 12]
```diff
@@ -4,18 +4,30 @@ SmartAnswer.isStartPage = function(slug) { // Used mostly during A/B testing
 }
 
 function linkToTemplatesOnGithub() {
-  $('*[data-debug-template-path]').each(function() {
+  var host = "https://github.com";
+  var organisation = "alphagov";
+  var repository = "smart-answers"
+  var branch = "deployed-to-production";
```
[the remark this answers is not in the document]
Comment (ikennaokpala) [reply app/assets/javascripts/helpers.js]: Well, I suppose this can be tackled.

File: app/assets/javascripts/helpers.js [hunk 13]
```diff
@@ -4,18 +4,30 @@ SmartAnswer.isStartPage = function(slug) { // Used mostly during A/B testing
...
+    var path = element.data("debug-partial-template-path");
+    var url = [host, organisation, repository, "blob", branch, path].join("/");
+    var filename = path.split("/").pop();
+    var anchor = $("<a>Partial template on GitHub</a>").attr("href", url).attr("style", "color: hotpink;").attr("title", filename);
     element.prepend(anchor);
-    element.attr('style', 'border: 3px solid deeppink; padding: 10px; margin: 3px');
-    element.removeAttr('data-debug-template-path');
+    element.attr("style", "border: 3px dotted hotpink; padding: 10px; margin: 3px");
+    element.removeAttr("data-debug-partial-template-path");
   });
```
[the remark this answers is not in the document]
Comment (ikennaokpala) [reply app/assets/javascripts/helpers.js]: Both still exist.. the context in mind is partials.

File: config/initializers/partial_renderer.rb [hunk 14]
```diff
@@ -0,0 +1 @@
+ActionView::PartialRenderer.prepend(SmartAnswerPartialTemplateWrapper)
```
[the remark this answers is not in the document]
Comment (ikennaokpala) [reply config/initializers/partial_renderer.rb]: Any other recommendation would be welcomed.

File: lib/smart_answer_partial_template_wrapper.rb (hunk 1 above)
Comment (ikennaokpala) [reply lib/smart_answer_partial_template_wrapper.rb]: I believe that this presents the opportunity to fine tune this..

File: lib/smart_answer_partial_template_wrapper.rb (hunk 3 above)
Comment (ikennaokpala) [reply lib/smart_answer_partial_template_wrapper.rb]: Yeah.. this is used as a filter to help identify and select partials that should be wrapped.

File: test/unit/smart_answer_partial_template_wrapper_test.rb [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+  attr_accessor :template
+
+  def render(*)
+    "<h1>Hello</h1>"
+  end
+end
+
+class SampleRenderer < TestRenderer
+  include SmartAnswerPartialTemplateWrapper
+end
```
[the remark this answers is not in the document]
Comment (ikennaokpala) [reply test/unit/smart_answer_partial_template_wrapper_test.rb]: Ideally yes

**Comment by ikennaokpala** [comment]
Rebasing from 3f46068 to 6f76ba5

**Comment by pmanrubia** [comment]
Thank you for your work @ikennaokpala, and for your review @floehopper. 
We are closing this PR, and we will keep it for reference.
